src/main/java/edu/umich/tl/analytics/CanvasCourseReportServlet.java

```diff
@@ -74,32 +84,121 @@ protected void doGet(HttpServletRequest request,HttpServletResponse response){
+			outputStream.close();
+		}else {
+			afterActionThreadIsDone(response, aCourseReportThread, outputStream);
+		}
+	}
+
+
+	private void afterActionThreadIsDone(HttpServletResponse response, Thread aCourseReportThread,
+			OutputStream outputStream)
+					throws IOException {
+		CourseReportTask courseReportTask = courseReportReqThreads.get(aCourseReportThread);
+		CoursesForSubaccounts cfs = courseReportTask.cfs;
+		if(cfs.subaccountCallHasErr || cfs.courseCallHasErr ) {
+			courseReportReqThreads.remove(aCourseReportThread);
+			response.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+			outputStream.write(props.getString("report.generation.err.msg").getBytes());
+			M_log.debug("Ending polling... Thread size: "+courseReportReqThreads.size());
+		}else {
+			response.setContentType("text/csv");
+			response.setHeader("Content-Disposition","attachment;filename=PublishedCanvasCoursesWithInstructorsFor"+courseReportTask.termName+".csv");
+			String outputResult = generateCSVFile(cfs).toString();
+			courseReportReqThreads.remove(aCourseReportThread);
+			outputStream.write(outputResult.getBytes());
+			outputStream.flush();
+			outputStream.close();
+			M_log.debug("Ending polling... Thread size: "+courseReportReqThreads.size());
+		}
+	}
 	private void getPublishedCourses(HttpServletRequest request,HttpServletResponse response) {
-		M_log.debug("getPublishedCourses: Called");
+		M_log.debug("getPublishedCourses(): Called");
 		String term = request.getParameter(TERM);
 		String termName = request.getParameter(TERM_NAME);
-		String subAccountUrl = canvasURL+"/api/v1/accounts/1/sub_accounts?recursive=true&per_page=100";
-		String url = canvasURL+"/api/v1/accounts/1/courses?enrollment_term_id="+term+"&published=true&per_page=100";
 		CoursesForSubaccounts cfs =new CoursesForSubaccounts();
-		getSubAccountInfo(subAccountUrl,cfs,response);
-		if(!cfs.subaccountCallHasErr) {
-			getThePublishedCourseList(url,cfs,response);
-			if(!cfs.courseCallHasErr) {
-				response.setContentType("text/csv");
-				response.setHeader("Content-Disposition","attachment;filename=PublishedCanvasCoursesWithInstructorsFor"+termName+".csv");
-				try {
-					OutputStream outputStream = response.getOutputStream();
-					String outputResult = generateCSVFile(cfs).toString();
-					outputStream.write(outputResult.getBytes());
-					outputStream.flush();
-					outputStream.close();
-				} catch (IOException e) {
-					M_log.error("Writing the csv file to the has some errror: ",e);
-				}
-			}
+		CourseReportTask courseReportTask=new CourseReportTask(term, termName,cfs);
+		Thread courseReportThread=new Thread(courseReportTask);
+		courseReportReqThreads.put(courseReportThread, courseReportTask);
+		courseReportThread.start();
+		PrintWriter out;
+		try {
+			out = response.getWriter();
+			out.print(courseReportThread.getId());
+			out.flush();
+		} catch (IOException e) {
+			e.printStackTrace();
 		}
 	}
 
+	public class CourseReportTask implements Runnable{
```

Comment:
Can you put this class in a separate file?  I'd prefer to see each class in its own file.  That will also help if you later write unit tests for the class.  I think they're easier to do if the class is in its own file.